### PR TITLE
Cleanup util.py, and test type hint check

### PR DESCRIPTION
**Pull request by math-fehr**
Remove deprecated functions in `util.py`, and improve typing in `util.py`.
Also, add tests for type hint checking

### CODE REVIEW

File: src/xdsl/util.py [hunk 1]
```diff
@@ -1,67 +1,59 @@
...
-from dataclasses import dataclass
-from xdsl.ir import Operation, SSAValue, BlockArgument, Block, Region, Attribute
+from inspect import isclass
+from typing import (Annotated, Any, TypeGuard, TypeVar, Union, cast, get_args,
+                    get_origin)
 
+# pyright: strict
 
-def new_op(op_name, num_results, num_operands, num_regions) -> type[Operation]:
+_T = TypeVar("_T")
```
[the remark this answers is not in the document]
Comment (math-fehr) [reply src/xdsl/util.py:7]: Would that makes things look necessarily better?
I'm used to not reuse type variables, but I actually don't know the best practices for this.
Also, how is this related to package import pollution? I don't think you need to import the type variables when you import a function using them?

**Comment by math-fehr** [comment]
Thanks for the code cleanup!
I'll remove the strict typing check though, since it should be for the entire project instead (our plan is to have everything satisfying the strict typing).

I am not sure we need tuples and sets, especially since this typing check is not often used in practice in our code. If you want to add it, could you add similar tests?

**Comment by K-W-Li** [comment]
Just realized that tuple can have different types and even ellipsis in typing...

OK, maybe remove tuple and set if there is no practical usage yet.

File: src/xdsl/util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (K-W-Li) [reply src/xdsl/util.py:5]: See [parser.py](https://github.com/xdslproject/xdsl/blob/6f5d6afd088184e6820d8040f14ef4367f4cbf06/src/xdsl/parser.py#L146) where the ```T``` is actually not used and ```parse_list```'s ```T``` indeed comes from another package due to ```from xdsl.dialects.builtin import *``` star import pollution.

[Related stack overflow question](https://stackoverflow.com/questions/47615480/best-practices-for-generic-functions-in-python) with comment 
> The only downside is that having that TypeVar on the outside does seem vaguely untidy/seems like it'll pollute the global namespace. You can partially mitigate this by renaming it to something like _T for example, to indicate it's private. You can also reuse this typevar whenever you need another generic class or function, which should also hopefully help keep the namespace pollution down.

To be honest, I don't know the best practice. Maybe #132 will alleviate the issue?

**Comment by K-W-Li** [comment]
> Thanks for the code cleanup! I'll remove the strict typing check though, since it should be for the entire project instead (our plan is to have everything satisfying the strict typing).
> 
> I am not sure we need tuples and sets, especially since this typing check is not often used in practice in our code. If you want to add it, could you add similar tests?

I was thinking about whether we need ```FrozenList```, from the current usage I think ```tuple``` can do the job.
That's why I add tuple support here just in case.

Correct me if I misunderstand some key feature in ```FrozenList``` usage.

File: src/xdsl/util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (math-fehr) [reply src/xdsl/util.py:5]: Yeah, this is clearly broken! Let's not share the `TypeVar` between files, which is why I made these ones private.
And yes, #132 will help us see where the problems are.

**Comment by math-fehr** [comment]
So, this function is only used on `builders` which help us simplify some code.
So we only use this function on the arguments of the builders, which are rarely tuples or sets (and also dictionaries, but I kept it because we already had it).
I think there is a very very low chance that we would use `FrozenList` in these builders, it is way more likely that we use lists in this case, or even tuples.
I just want to avoid passing time adding this code (especially for the tests) for something that we might never use in practice.

**Pull request merged**